### PR TITLE
[RELAY] Basic block normal form

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

As mentioned in https://discuss.tvm.ai/t/basic-block-normal-form/5908/5, the requirements for basic block normal form are:
- D0, Define a block as a group of expressions implied by the scope structure.
- D1: Each graph node can only belong to a single block
- D2: For values that are being used in multiple blocks, they have to be referred by a Var which is defined in a block

We rely on `analysis::DependencyGraph` to compute the dependency among nodes, and use LCA to compute the scope of each node. As we traverse the dependency graph, we record the expression with a "lifted scope" if its original scope is different from the result of LCA of nodes depending on it. For instance, a variable shared by the true_branch and false_branch in an `If` statement. If any expression with lifted scope is found, the function violates the basic block normal form. 

The expressions with lifted scope need to be bound by let binding. It is done with a pass almost the same as the ANF pass. Same as ANF we use a let list per scope to help construct let bindings of expressions. The main difference with ANF pass is that we only push expressions to the let list if the expression's scope is lifted, otherwise we return the original expression. When we come across If/function node, we get the expression generated from the let list of the corresponding scope. 

@MarisaKirisame would you suggest we merge the two passes into a single class?

### CODE REVIEW

**Comment by jroesch** [comment]
cc @MarisaKirisame

**Comment by tqchen** [comment]
Thanks @eric-haibin-lin , it would also be useful to have a verification analysis util that verifies the IR is in the basic block form

**Comment by ANSHUMAN87** [comment]
@eric-haibin-lin : Thanks for the PR! I believe this feature will be really helpful for certain detection & mutation process.
Please let me know, once your PR is ready for review. Would like to get a hold on it!

**Comment by tqchen** [comment]
cc @MarisaKirisame @mbrookhart @electriclilies @jroesch would be great if you can help to take a look  at the PR

**Comment by junrushao1994** [comment]
Ping me if the PR is ready to review :-)

File: src/relay/transforms/to_a_normal_form.cc [hunk 1]
```diff
@@ -50,10 +50,17 @@ Scope LCA(Scope lhs, Scope rhs) {
   return lhs;
 }
 
-std::unordered_map<DependencyGraph::Node*, Scope> CalcScope(const DependencyGraph& dg) {
-  std::unordered_map<DependencyGraph::Node*, Scope> expr_scope;
+// return a set of Exprs whose scope should be lifted to due dependencies.
+void CalcScope(const DependencyGraph& dg,
+               std::unordered_map<DependencyGraph::Node*, Scope>* expr_scope,
```
Comment (MarisaKirisame) [inline src/relay/transforms/to_a_normal_form.cc:55]: Why this change?

File: src/relay/transforms/to_basic_block_normal_form.cc [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+  using ExprSet = std::unordered_set<Expr, ObjectPtrHash, ObjectPtrEqual>;
+
+ public:
+  static Expr ToBasicBlockNormalForm(const Expr& e, const DependencyGraph& dg,
+                                     std::unordered_map<DependencyGraph::Node*, Scope>* node_scope,
+                                     ExprSet* lifted) {
+    FillBasicBlock fi(dg, node_scope, lifted);
+    auto var = fi.VisitExpr(e);
+    auto scope = fi.GetScope(e);
+    auto ret = scope->ll->Get(var);
```
Comment (junrushao1994) [inline src/relay/transforms/to_basic_block_normal_form.cc:52]: nit: use less auto, unless the type can be seen in the same line

File: src/relay/transforms/to_basic_block_normal_form.cc [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+                 std::unordered_set<Expr, ObjectPtrHash, ObjectPtrEqual>* lifted)
+      : dg_(dg), node_scope_(node_scope), lifted_(lifted) {}
+
+  Scope GetScope(const Expr& e) { return node_scope_->at(dg_.expr_node.at(e)); }
+
+  Scope GetSubScope(const Expr& e, size_t i) {
+    DependencyGraph::Node* n = dg_.expr_node.at(e);
+    auto h = n->children.head;
+    while (i != 0) {
+      CHECK(h);
```
Comment (junrushao1994) [inline src/relay/transforms/to_basic_block_normal_form.cc:73]: maybe add more error msg?

File: src/relay/transforms/to_basic_block_normal_form.cc [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+  }
+
+  Expr VisitExpr(const Expr& e, const Var& v) final {
+    if (memo.count(e) == 0) {
+      memo.insert({e, ExprFunctor<Expr(const Expr&, const Var&)>::VisitExpr(e, v)});
+    } else if (v.defined()) {
+      GetScope(e)->ll->Push(v, memo.at(e));
+    }
+    auto ret = memo.at(e);
+    return ret;
```
Comment (junrushao1994) [inline src/relay/transforms/to_basic_block_normal_form.cc:88]: nit
```suggestion
    return memo.at(e);
```

**Comment by eric-haibin-lin** [comment]
I added a short description in the PR summary and an analysis pass `check_basic_block_normal_form`. The PR is ready for review now, thanks!

File: src/relay/transforms/pass_util.h [hunk 5]
```diff
@@ -184,6 +188,37 @@ struct TreeBranchNode : TreeNode<ConditionObjectPtr> {
...
+};
+
+/*! \brief Calculate the scope of nodes in the dependency graph by least common ancestor.
+ *
+ *  \param dg the input dependency graph
+ *  \param expr_scope the output node -> scope mapping for all nodes.
+ *  \param lifted_exprs the output set of expressions whose scope is lifted due to dependency
+ */
+void CalcScope(const DependencyGraph& dg,
+               std::unordered_map<DependencyGraph::Node*, Scope>* expr_scope,
```
Comment (MarisaKirisame) [inline src/relay/transforms/pass_util.h:215]: why use pointer instead of returning?

Comment (eric-haibin-lin) [reply src/relay/transforms/pass_util.h:215]: @MarisaKirisame thanks for the review! For BBlock I want to get a set of expressions whose scope should be lifted and rewritten with let bindings. I found `CalcScope` already performs finding the LCA for all nodes in the dependency graph, so i just added a few lines of code so that the set of corresponding expressions are also returned by this function. However, setting the return type to `std::pair<std::unordered_map<node, scope> std::unordered_set<expr>>` causes cpplint to fails due to the long length. So I instead used pointers for both outputs. Please let me know the preferred way for returning both outputs. Thanks!

Comment (MarisaKirisame) [reply src/relay/transforms/pass_util.h:215]: I think we should disable cppint and use tuple, or just do a few typedef to shorten stuff.

Comment (eric-haibin-lin) [reply src/relay/transforms/pass_util.h:215]: Ok I'll update the code

File: src/relay/transforms/to_basic_block_normal_form.cc [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+#include "let_list.h"
+#include "pass_util.h"
+
+namespace tvm {
+namespace relay {
+
+/* Fill expressions based on each scope's let list. Different from FillANF,
+ * only expressions with lifted scope will be pushed to the let list.
+ */
+class FillBasicBlock : ExprFunctor<Expr(const Expr&, const Var&)> {
```
Comment (MarisaKirisame) [inline src/relay/transforms/to_basic_block_normal_form.cc:42]: can you reuse the Fill in to_anf? the code is almost identically the same, and duplicating it will make maintenance significantly harder.

File: src/relay/transforms/pass_util.h [hunk 7]
```diff
@@ -184,6 +189,89 @@ struct TreeBranchNode : TreeNode<ConditionObjectPtr> {
...
+ * If it is not null, Fill is required to bind the transformed result to that var.
+ */
+class Fill : ExprFunctor<Expr(const Expr&, const Var&)> {
+ public:
+  static Expr ToANormalForm(const Expr& e, const DependencyGraph& dg, NodeScopeMap* node_scope);
+
+  // For basic block normal form, bind expressions only if the original expression's
+  // scope should be lifted
+  static Expr ToBasicBlockNormalForm(const Expr& e, const DependencyGraph& dg,
+                                     NodeScopeMap* node_scope, ExprSet* lifted);
```
Comment (mbrookhart) [inline src/relay/transforms/pass_util.h:234]: I really like that you're reusing the infrastructure from A-Normal form to do this, there's a lot of overlap. I don't love that this class violates the Single Responsiblity Principle and that you have to introduce branches in a couple of the recursive functions to handle the two different cases. Given that this class has a private constructor, I'm not sure it 100% matters, but there might  be a cleaner solution?

File: src/relay/transforms/to_a_normal_form.cc [hunk 8]
```diff
@@ -81,171 +70,187 @@ std::unordered_map<DependencyGraph::Node*, Scope> CalcScope(const DependencyGrap
...
+}
 
-  Expr VisitExpr_(const CallNode* c, const Var& v) final {
-    Expr e = GetRef<Expr>(c);
-    std::vector<Expr> args;
-    for (const auto& a : c->args) {
-      args.push_back(VisitExpr(a));
-    }
-    return Compound(e, Call(VisitExpr(c->op), args, c->attrs, c->type_args), v);
+Expr Fill::VisitExpr_(const CallNode* c, const Var& v) {
```
Comment (mbrookhart) [inline src/relay/transforms/to_a_normal_form.cc:155]: Why remove `final` from all of these functions?

File: src/relay/transforms/to_basic_block_normal_form.cc [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+#include "../../support/arena.h"
+#include "../analysis/dependency_graph.h"
+#include "let_list.h"
+#include "pass_util.h"
+
+namespace tvm {
+namespace relay {
+
+Expr ToBasicBlockNormalFormAux(const Expr& e) {
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply src/relay/transforms/to_basic_block_normal_form.cc:39]: I guess it's because of https://en.wikipedia.org/wiki/Basic_block  ?

File: src/relay/transforms/to_a_normal_form.cc (hunk 8 above)
Comment (eric-haibin-lin) [reply src/relay/transforms/to_a_normal_form.cc:155]: `final` is still used in the header at declaration https://en.cppreference.com/w/cpp/language/final#Syntax

File: src/relay/transforms/pass_util.h (hunk 7 above)
Comment (eric-haibin-lin) [reply src/relay/transforms/pass_util.h:234]: thanks for bringing this up. I was also thinking about how to best leverage the existing code without too much ad-hoc code or duplication. I think adding an inclusion/exclusion argument to `Fill` is still acceptable, and the Fill constructor is private anyway

File: src/relay/transforms/to_basic_block_normal_form.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (mbrookhart) [reply src/relay/transforms/to_basic_block_normal_form.cc:39]: :) I missed this naming influence in the conversations @tqchen and I had a few months ago. Thanks! That makes things a lot clearer for me.

**Comment by MarisaKirisame** [comment]
Thanks @eric-haibin-lin @mbrookhart @junrushao1994

**Comment by MarisaKirisame** [comment]
I dont think it break single responsibility - the code is doing conversion and a single configuration param denote which conversion it is. Another way to think about it is that that visitor is declaring a 'scoped mutator' and toanf/tobbnf is two subclass of it.
We can refactor to the latter later if needed.

**Comment by mbrookhart** [comment]
> I dont think it break single responsibility - the code is doing conversion and a single configuration param denote which conversion it is. Another way to think about it is that that visitor is declaring a 'scoped mutator' and toanf/tobbnf is two subclass of it.
> We can refactor to the latter later if needed.

That is exactly breaking the SRP :) The SRP-compliant way to do this would be exactly what you said, a base class and two subclasses. That being said, since we're not allowing anyone to inherit from this class, and the constructor is private, it's probably not the end of the world, I'm cool with leaving it the way it is.

**Comment by MarisaKirisame** [comment]
We can allow it if needed!

**Comment by MarisaKirisame** [comment]
@eric-haibin-lin please at me when it is done. I will merge it then.

**Comment by MarisaKirisame** [comment]
Thanks @eric-haibin-lin @mbrookhart @junrushao1994
